sdks/java/core/src/main/java/org/apache/beam/sdk/coders/RowCoder.java

```diff
@@ -17,69 +17,146 @@
+      .build();
+
+  private static final Map<TypeName, Integer> ESTIMATED_FIELD_SIZES =
+      ImmutableMap.<TypeName, Integer>builder()
+          .put(TypeName.BYTE, Byte.BYTES)
+          .put(TypeName.INT16, Short.BYTES)
+          .put(TypeName.INT32, Integer.BYTES)
+          .put(TypeName.INT64, Long.BYTES)
+          .put(TypeName.FLOAT, Float.BYTES)
+          .put(TypeName.DOUBLE, Double.BYTES)
+          .put(TypeName.DECIMAL, 32)
+          .put(TypeName.BOOLEAN, 1)
+          .put(TypeName.DATETIME, Long.BYTES)
+          .build();
+
   private static final BitSetCoder nullListCoder = BitSetCoder.of();
 
-  private RowType rowType;
-  private List<Coder> coders;
+  private Schema schema;
+
+  /**
+   * Returns the coder used for a given primitive type.
+   */
+  public static <T> Coder<T> coderForPrimitiveType(TypeName typeName) {
+    return (Coder<T>) CODER_MAP.get(typeName);
+  }
+
+  /**
+   * Return the estimated serialized size of a give row object.
+   */
+  public static long estimatedSizeBytes(Row row) {
+    Schema schema = row.getSchema();
+    int fieldCount = schema.getFieldCount();
+    int bitmapSize = (((fieldCount - 1) >> 6) + 1) * 8;
 
-  private RowCoder(RowType rowType, List<Coder> coders) {
-    this.rowType = rowType;
-    this.coders = coders;
+    int fieldsSize = 0;
+    for (int i = 0; i < schema.getFieldCount(); ++i) {
+      fieldsSize += estimatedSizeBytes(schema.getField(i).getTypeDescriptor(), row.getValue(i));
+    }
+    return bitmapSize + fieldsSize;
   }
 
-  public static RowCoder of(RowType rowType, List<Coder> coderArray) {
-    if (rowType.getFieldCount() != coderArray.size()) {
-      throw new IllegalArgumentException("Coder size doesn't match with field size");
+  private static long estimatedSizeBytes(FieldTypeDescriptor typeDescriptor, Object value) {
+    switch (typeDescriptor.getType()) {
+      case ROW:
+        return estimatedSizeBytes((Row) value);
+      case ARRAY:
+        List list = (List) value;
+        long listSizeBytes = 0;
+        for (Object elem : list) {
+          listSizeBytes += estimatedSizeBytes(typeDescriptor.getComponentType(), elem);
+        }
+        return 4 + listSizeBytes;
+      case STRING:
+        // Not always accurate - String.getBytes().length() would be more accurate here, but slower.
+        return ((String) value).length();
+      default:
+        return ESTIMATED_FIELD_SIZES.get(typeDescriptor.getType());
     }
-    return new RowCoder(rowType, coderArray);
   }
 
-  public RowType getRowType() {
-    return rowType;
+  private RowCoder(Schema schema) {
+    this.schema = schema;
+  }
+
+  public static RowCoder of(Schema schema) {
+    return new RowCoder(schema);
+  }
+
+  public Schema getSchema() {
+    return schema;
+  }
+
+  Coder getCoder(FieldTypeDescriptor fieldTypeDescriptor) {
+    if (TypeName.ARRAY.equals(fieldTypeDescriptor.getType())) {
+      return ListCoder.of(getCoder(fieldTypeDescriptor.getComponentType()));
```

Comment:
we probably should still keep the `Map<FieldTypeDescriptor, Coder>` to cache everything